crates/ide/src/fetch_crates.rs

```diff
@@ -0,0 +1,59 @@
+use ide_db::{
+    base_db::{CrateOrigin, SourceDatabase, SourceDatabaseExt},
+    RootDatabase,
+};
+
+#[derive(Debug)]
+pub struct CrateInfo {
+    pub name: String,
+    pub version: String,
+    pub path: String,
+}
+
+pub(crate) fn fetch_crates(db: &RootDatabase) -> Vec<CrateInfo> {
+    let crate_graph = db.crate_graph();
+    crate_graph
+        .iter()
+        .map(|crate_id| &crate_graph[crate_id])
+        .filter(|&data| !matches!(data.origin, CrateOrigin::Local { .. }))
+        .map(|data| {
+            let crate_name = crate_name(data);
+            let version = data.version.clone().unwrap_or_else(|| "".to_owned());
+            let crate_path = crate_path(db, data, &crate_name);
+
+            CrateInfo { name: crate_name, version, path: crate_path }
+        })
+        .collect()
+}
+
+fn crate_name(data: &ide_db::base_db::CrateData) -> String {
+    data.display_name
+        .clone()
+        .map(|it| it.canonical_name().to_owned())
+        .unwrap_or("unknown".to_string())
+}
+
```

Comment:
I feel that the function below is SUPER hacky, but I couldn't figure another way around it.